Sources/NIO/EventLoop.swift

```diff
@@ -952,33 +981,76 @@ internal final class SelectableEventLoop: EventLoop {
         try self.selector.close()
     }
 
-    fileprivate func close0() throws {
-        if inEventLoop {
-            self.lifecycleState = .closed
-        } else {
-            self.execute {
-                self.lifecycleState = .closed
+    internal func initiateClose(queue: DispatchQueue, completionHandler: @escaping (Result<Void, Error>) -> Void) {
+        func doClose() {
+            self.assertInEventLoop()
+            // There should only ever be one call into this function so we need to be up and running, ...
+            assert(self.internalState == .runningAndAcceptingNewRegistrations)
+            self.internalState = .runningButNotAcceptingNewRegistrations
+
+            self.externalStateLock.withLock {
+                // ... but before this call happened, the lifecycle state should have been changed on some other thread.
+                assert(self.externalState == .closing)
             }
-        }
-    }
 
-    /// Gently close this `SelectableEventLoop` which means we will close all `SelectableChannel`s before finally close this `SelectableEventLoop` as well.
-    public func closeGently() -> EventLoopFuture<Void> {
-        func closeGently0() -> EventLoopFuture<Void> {
-            guard self.lifecycleState == .open else {
-                return self.makeFailedFuture(EventLoopError.shutdown)
+            self.selector.closeGently(eventLoop: self).whenComplete { result in
+                self.assertInEventLoop()
+                assert(self.internalState == .runningButNotAcceptingNewRegistrations)
+                self.internalState = .noLongerRunning
+                self.execute {}
```

Comment:
What's this for?

Reply:
adding a comment: this is to force the event loop to create a new tick and then realise that it is no longer running.